lib/libmuser.c

```diff
@@ -251,7 +223,38 @@ init_sock(lm_ctx_t *lm_ctx)
 static int
 open_sock(lm_ctx_t *lm_ctx)
 {
-    return accept(lm_ctx->fd, NULL, NULL);
+    int ret, fd;
+    struct vfio_user_request cmd = {};
+
+    fd = accept(lm_ctx->fd, NULL, NULL);
+    if (fd < 0) {
+	return fd;
+    }
+
+    /* Send a header VFIO_USER_VERSION message */
+    cmd.hdr.command = VFIO_USER_VERSION;
+    cmd.hdr.msg_size = sizeof(struct vfio_user_header);
+    ret = write(fd, &cmd, sizeof(struct vfio_user_header));
+    if (ret < 0) {
+        lm_log(lm_ctx, LM_ERR, "failed to send VFIO_USER_VERSION\n");
+        return -EFAULT;
+    }
+    memset(&cmd, 0, sizeof(cmd));
+
+retry:
+    /* TODO: just use a header now */
+    ret = recv(fd, &cmd, sizeof(struct vfio_user_header), lm_ctx->sock_flags);
+    if (ret < 0) {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+		goto retry;
+	}
+        lm_log(lm_ctx, LM_ERR, "failed to recv VFIO_USER_VERSION\n");
+        return -EFAULT;
+    }
+    assert(cmd.hdr.command == VFIO_USER_VERSION);
+    assert(cmd.hdr.flags.type == VFIO_USER_MESSAGE_REPLY);
+
```

Comment:
We should be exchanging a JSON obejct where we check that the version is the same etc.